nixos/modules/services/matrix/mautrix-meta.nix

```diff
@@ -0,0 +1,489 @@
+            old_umask=$(umask)
+            umask 0177
+            ${lib.getExe pkgs.envsubst} \
+              -o '${settingsFile cfg}' \
+              -i '${settingsFileUnsubstituted cfg}'
+
+            umask 0127
+
+            has_tokens="$(${lib.getExe pkgs.yq} '.appservice | has("as_token") and has("hs_token")' '${settingsFileUnsubstituted cfg}')"
+            generated=0
+
+            if [[ (! -f '${cfg.registrationFile}') ]]; then
+              echo "Generating registration file"
+              ${lib.getExe upperCfg.package} \
+                --generate-registration \
+                --config='${settingsFile cfg}' \
+                --registration='${cfg.registrationFile}'
+              generated=1
+            fi
+
+            umask 0177
+
+            # Make sure correct tokens are in the registration file
+            if [[ "$has_tokens" == "true" ]]; then
+              echo "Substituting as_token, hs_token inside the registration file"
+              ${lib.getExe pkgs.yq} -s '.[1].as_token = .[0].appservice.as_token
+                | .[1].hs_token = .[0].appservice.hs_token
+                | .[1]' '${settingsFileUnsubstituted cfg}' '${cfg.registrationFile}' \
+                > '${cfg.registrationFile}.tmp'
+              mv '${cfg.registrationFile}.tmp' '${cfg.registrationFile}'
+            fi
+
+            # Generation overrides as_token, hs_token.
+            if [[ "$generated" == "1" || "$has_tokens" == "false" ]]; then
+              echo "Substituting as_token, hs_token inside the config file"
+              ${lib.getExe pkgs.yq} -s '.[0].appservice.as_token = .[1].as_token
+                | .[0].appservice.hs_token = .[1].hs_token
+                | .[0]' '${settingsFile cfg}' '${cfg.registrationFile}' \
+                > '${settingsFile cfg}.tmp'
+              mv '${settingsFile cfg}.tmp' '${settingsFile cfg}'
+            fi
+
+            umask $old_umask
+          ''; #+ lib.optionalString config.services.matrix-synapse.enable ''
+          #   chown mautrix-meta:matrix-synapse '${cfg.registrationFile}'
+          # '';
+
+          serviceConfig = {
+            UMask = 0027;
+
+            User = "mautrix-meta";
+            Group = "mautrix-meta";
+
+            SystemCallFilter = [ "@system-service" "@chown" ];
+            CapabilityBoundingSet = [ "CAP_CHOWN" ];
+
+            ProtectSystem = "strict";
+            ProtectHome = true;
+
+            StateDirectory = cfg.dataDir;
+            ReadWritePaths = fullDataDir cfg;
+          };
+
+          restartTriggers = [ (settingsFileUnsubstituted cfg) ];
+        };
+          "${metaName name}" = {
+            description = "Mautrix-Meta bridge - ${metaName name}";
+            wantedBy = [ "multi-user.target" ];
+            wants = [ "network-online.target" ] ++ cfg.serviceDependencies;
+            after = [ "network-online.target" ] ++ cfg.serviceDependencies;
+
+            environment.HOME = fullDataDir cfg;
```

Comment:
If we are doing this is is probably best to just configure the user's home directory.

Reply:
i've tested with postgresql and it doesn't seem needed. This was in mautrix-telegram with comment that it's needed when postgresql is used as the bridge tries to create something in HOME.